### ISSUE TITLE
Add name to Class Parameter()

### ISSUE THREAD

**Issue by zzzwen** [pytorch/pytorch#37554]
## 🚀 Feature
Add name to Class Parameter() [link] (https://github.com/pytorch/pytorch/blob/master/torch/nn/parameter.py).

## Motivation
Currently, parameter names are available via `nn.Module.name_parameter()`, it is good enough for a model that locates on a single machine. However, once we start to pass parameters around via RPC, a stable name inside `class Parameter()`  become really handy. 
A stable name facilitate a lot of distributed use cases, say we need to save a distributed model "distributedly", an obvious key to look up a parameter (for offline manipulation or restoring) is the parameter name. And debugging a distributed model will be easier too.

There are couple requirement for this name 
1. The name should be stable. If one instantiates the same model class multiple times, the name of a particle parameter should be exactly the same.
2. The name should be readable. 
3. The name might need to reflect a certain set (not all) of operation on the parameter. e.g. a `detach()` parameter will have the keyword "detached" in it. 

P.S. I did a quick search and did not see an issue for this, feel free to dupe it if there is an open one already.

**Comment by mrshenli** [comment]
Adding name to `Parameter` is reasonable request to me. @gchanan @ezyang does this request make sense to you?

I am removing `distributed` and `rpc` tags and adding `nn` tag as this is a FR to `torch.nn`.

> There are couple requirement for this name
> 
> 1. The name should be stable. If one instantiates the same model class multiple times, the name of a particle parameter should be exactly the same.
> 2. The name should be readable.
> 3. The name might need to reflect a certain set (not all) of operation on the parameter. e.g. a detach() parameter will have the keyword "detached" in it.

If we follow today's `nn.Module`'s design, it is application's responsibility to meet the above requirement, as the `Parameter` name is basically the attribute name:

https://github.com/pytorch/pytorch/blob/ccab142197d4647e397fc492ba4e7d857fa07a6d/torch/nn/modules/module.py#L623-L644

**Comment by mrshenli** [comment]
Adding this issue to triage review to discuss whether we should introduce this change.

**Comment by albanD** [comment]
This might encounter some issues similar to https://github.com/pytorch/pytorch/pull/38604 (that did not go in) where most code currently assumes that Parameters are just Tensors and nothing else.
So it will be significant work to properly add any new attribute to the nn.Parameter object if we want them to remain attached when moved across devices, serialized, used in jit, etc.

**Comment by ezyang** [comment]
> The name might need to reflect a certain set (not all) of operation on the parameter. e.g. a detach() parameter will have the keyword "detached" in it.

This seems like an actually pretty crazy requirement. Once you detach a Parameter it is no longer a Parameter, it's just a good old Tensor.

> A stable name facilitate a lot of distributed use cases, say we need to save a distributed model "distributedly", an obvious key to look up a parameter (for offline manipulation or restoring) is the parameter name. And debugging a distributed model will be easier too.

Could you tell us more about this use cases?

**Comment by zzzwen** [comment]
To make the terminology correct, this request should actually be "add a stable/readable name to torch.Tensor" instead of nn.Parameter. 

Basically, if a Parameter is referred outside of a nn.Module (at this point, it is just a plain Tensor?), I do not have a simple way to know where it was from.

e.g. without walking the module, I won't be able to figure out what's the key in optim status.
https://github.com/pytorch/pytorch/blob/6e2bb1c05442010aff90b413e21fce99f0393727/torch/optim/adagrad.py#L42

**Comment by mruberry** [comment]
I think we're going to need a more thorough design proposal to review this issue. 

You DO have a way to know which module(s) a tensor is associated with. You can, for example, create a tensor id -> module map if you like. So how would a name vs the tensor id work?

**Comment by ezyang** [comment]
> I do not have a simple way to know where it was from.

Could you tell us in more detail why it is necessary to know where it is from? In particular, what you meant by "offline manipulation or restoring"?

**Comment by JamesSand** [comment]
> Could you tell us in more detail why it is necessary to know where it is from? In particular, what you meant by "offline manipulation or restoring"?

 @ezyang  Hi! I have met quite same issue. I am modifying the optimizer state in AdamW. So I need to visualize some property in parameters grad. I have to determine which module does each parameter correspond to... Therefore I need a name attribute in nn.Parameters

**Comment by albanD** [comment]
Hey!

Here is a way to achieve this:
```python
import torch
from torch import nn

mod = nn.Sequential(nn.Linear(1, 1), nn.Conv2d(2, 2, 2, 2))
opt = torch.optim.SGD(mod.parameters())

param_to_name = {p: name for name, p in mod.named_parameters()}

some_param_from_optimizer = opt.param_groups[0]["params"][0]
print(param_to_name[some_param_from_optimizer])
```

Note that this will only give you one of the names if a single parameter is shared within your network though. And I don't think we have a good solution to handle this case as this single Parameter would now have two names.
